Fixes #26772 - Use react-router for hw models page

xprazak2 opened the pull request:
Using react-router on HW models page.

Main changes:
* Using the PageLayout component, added minor visual tweaks
* using LoadingState on initial load
* using react-router 

TODO:
* searching does not work
* add empty state when no hw models found
* tests
* see if state structure can be improved
* deleting an item refreshes page - fix separately

cc @Laviro, @glekner 

<!--
Simple description of what is fixed/introduced.

Prerequisites for testing:
* VMware cluster
* Two smart proxies

Tests needed (anticipated impacts):
* Hosts list - mainly power column
* Power status on host page
-->

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [Translating section in the guide]
* Suggest prerequisites for testing and testing scenarios following example above.
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

-->

glekner commented on package.json:
```diff
@@ -96,6 +96,7 @@
     "patternfly-react": "^2.34.1",
     "patternfly-react-extensions": "^2.18.8",
     "prop-types": "^15.6.0",
+    "query-string": "^6.5.0",
```
we have `urlHelpers.js` to avoid adding this package

glekner commented on webpack/assets/javascripts/react_app/components/LoadingState/index.js:
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
+import { LoadingState as PfLoadingState } from 'patternfly-react';
```
I don't like this component, its throwing errors for me, will add a default `LoadingComponent` in #6767

glekner commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js:
```diff
@@ -0,0 +1,33 @@
+    const { loading, loadingError } = this.props;
+
+    return (
+      <LoadingState loading={loading}>
```
with `HOC.js` there won't be a need for conditional rendering inside the Page. also added in #6767

glekner commented on webpack/assets/javascripts/react_app/routes/Models/index.js:
```diff
@@ -0,0 +1,7 @@
+import React from 'react';
+import ModelsPage from './ModelsPage';
+
+export default {
```
this is a great start, I guess we would like to add more routes here in the future (new ,delete, edit)

glekner reviewed (commented):
Thanks for this @xprazak2, looks great.
We should avoid some patterns I see here once #6767 is merged.

I think its best I will host a deep dive on this whole react-router topic, so we can all agree on the structure and patterns.

xprazak2 commented:
I rebased on top of latest additions to react-router infrastructure.

* I removed the LoadingState component
* using `withRenderHandler` and `callOnMount` from `HOC` with default guard components
* using our custom `urlHelpers` instead of `query-string`

ekohl commented on webpack/assets/javascripts/react_app/common/urlHelpers.js:
```diff
@@ -28,7 +28,7 @@ export const getURIpage = () => Number(getURI().query(true).page) || 1;
  * Get updated perPage param
  */
-export const getURIperPage = () => Number(getURI().query(true).per_page);
+export const getURIperPage = () => Number(getURI().query(true).per_page) || 25;
```
We have the `entries_per_page` setting. Perhaps this should honor it? It also defaults to 20 so if you don't honor it, matching defaults might be a good thing.

ohadlevy replied:
Don't we have this metadata now as part of the app component?

ekohl commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPageActions.js:
```diff
@@ -0,0 +1,33 @@
+} from '../consts';
+
+export const loadInitialModels = (query = {}) => dispatch => {
+  const url = new URI(`/api/${controller}`);
```
Please use the [foremanUrl](https://github.com/theforeman/foreman/blob/c5fbd28993fbe29f3457a329245e8975dd4c7667/webpack/assets/javascripts/foreman_tools.js#L130-L133) helper for URLs

ekohl commented on webpack/assets/javascripts/react_app/routes/Models/consts.js:
```diff
@@ -0,0 +1,19 @@
+  bookmarks: {
+    url: "/api/bookmarks",
+    canCreate: true,
+    documentationUrl: "https://theforeman.org/manuals/1.23/index.html#4.1.5Searching"
```
Somewhere else we just use the relative part. Hopefully that will then refer to the correct version.
https://github.com/theforeman/foreman/blob/c5fbd28993fbe29f3457a329245e8975dd4c7667/webpack/assets/javascripts/react_app/constants.js#L22

ekohl commented on webpack/assets/javascripts/react_app/routes/Models/consts.js:
```diff
@@ -0,0 +1,19 @@
+    useKeyShortcuts: true
+  },
+  bookmarks: {
+    url: "/api/bookmarks",
```
Should this also use the URL helper?

xprazak2 replied:
This gets passed only to search component, so I think it is ok as is.

glekner commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/index.js:
```diff
@@ -0,0 +1,38 @@
+
+export default compose(
+  connect(mapStateToProps, mapDispatchToProps),
+  callOnMount(({ loadInitialModels }) => loadInitialModels(getParams()))
```
🚀

glekner commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js:
```diff
@@ -0,0 +1,23 @@
+);
+
+export default withRenderHandler({
+  Component: ModelsPage
```
Looks so clean now, awesome!

one issue - `withRenderHandler` should wrap `ModelsTable` and not the whole `PageLayout`,
so when an error/empty/loading occurs, the user can still see the PageLayout ( header, search, breadcrumbs etc)

glekner commented on webpack/assets/javascripts/react_app/routes/routes.js:
```diff
@@ -1,3 +1,4 @@
+import Models from './Models';
 
-export const routes = [HostWizard];
+export const routes = [HostWizard, ...Models];
```
just so i understand, the spread is for future nested routing? e.g. `/models/new`

xprazak2 replied:
Yes, that is the intention.

glekner reviewed (commented):
Great progress so far, I'm also working on a generic reducer so we can reuse reducer logic on API calls, will update on it.

xprazak2 commented:
Rebased on changes in develop and added a couple of changes:

* using common page reducer
* loading component does not wrap page layout
* perPage from metadata as a default value

glekner commented on webpack/assets/javascripts/react_app/ReactApp/ReactApp.js:
```diff
@@ -2,14 +2,13 @@ import React from 'react';
 import AppSwitcher from '../routes';
 
-const ReactApp = ({ data: { layout } }) => (
+const ReactApp = ({ data: { layout, metadata } }) => (
```
we decided to avoid this prop drilling, [metadata is in the redux state](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/ReactApp/index.js#L25) so you should use selectors.

glekner commented on webpack/assets/javascripts/react_app/common/HOC.js:
```diff
@@ -43,3 +45,10 @@ export const withRenderHandler = ({
 };
+
+
+export const withPageLayout = layoutProps => WrappedComponent => componentProps => (
```
I like this, but what about functions that need props?

e.g.

```js
component = ({ neededProp }) => <PageLayout onSearch={() => callback(neededProp)} />
```

glekner commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTable.js:
```diff
@@ -34,13 +42,33 @@ const ModelsTable = ({
     );
   }
 
+  const onPageChange = (history) => (paginationArgs) => {
```
we should move this to the actions file.

glekner commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js:
```diff
@@ -0,0 +1,28 @@
+
+import searchProps from '../consts';
+
+const layoutProps = {
```
whats happening when I search? Turbolinks?
We should trigger an API call whenever the query changes, and update the table. its easy because we can handle the history stack. I implemented it all in #6733

glekner commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js:
```diff
@@ -0,0 +1,28 @@
+  })(componentProps);
+
+export default compose(
+  withPageLayout(layoutProps),
```
I like this, still there is an issue with missing props as stated above.

glekner commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/index.js:
```diff
@@ -0,0 +1,36 @@
+    isLoading: modelsAreLoading(state),
+    modelsMessage: modelsMessage(state),
+    hasError: modelsHaveError(state),
+    hasData: modelsHaveData(state),
```
you can add `metadata` here, since its in the redux state, and it will be available in `callOnMount`

glekner commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -4,19 +4,21 @@ import { routes } from './routes';
 let currentLocation = null;
 
-const AppSwitcher = () => (
+const AppSwitcher = ({ metadata }) => (
```
as stated above, metadata is now in the redux state so we should avoid this state (prop drilling)

glekner reviewed (changes requested):
Great work @xprazak2 🥇

ohadlevy commented on app/views/models/welcome.html.erb:
```diff
@@ -1,11 +0,0 @@
-<% content_for(:title, _("Hardware Models")) %>
```
FYI, this will need to undo the changes at https://github.com/theforeman/foreman/pull/6818/ as well.

Commit:
Fixes #26772 - Use react-router on hw models page

xprazak2 commented:
I had a bit of time to continue with this. I rebased on top of the work that has been done for Audits page.

xprazak2 commented:
There is a small problem with a toast when a new model is created - it does not display after redirect to models page, but is displayed when we go to /models/new again. 

Probably the best way to resolve this is to have new page in react as well...

jeremylenz reviewed (changes requested):
Loader looks better!

Some things seem to be not working:

* When I click an existing model to go to the Edit screen, the URL in the URL bar changes, but I remain on the same page.
* After deleting a model, I am redirected to the `/new` page - I would expect to be redirected to the index instead.  (This issue may be preexisting)
* When I click Cancel on the `/new` form, the URL changes back to `/models` but I see a blank page.
* Location changes seem to still be handled by Turbolinks and not rendered immediately by React.
* When viewing the tree in the React dev tools, the `<Route />` doesn't seem to have any children.  I would expect the child of the Route to be the component you want to display.  (This issue seems to be preexisting as well - see `react_app/routes/index.js` line 38 where the `render` prop returns `null`.  @sharvit any idea why it's like this?)

xprazak2 commented:
I fixed navigating to edit page and the cancel button now behaves correctly. I also modified the cancel button to use react-router.

Regarding the missing children, I think it is expected in cases when react-router is unable to match a route - then turbolinks come into play. On edit/new page, I see `Route` without children:

![comp-tree](https://user-images.githubusercontent.com/2664090/67947452-fb133a00-fbe3-11e9-87e1-81fc06e4a54c.png)

On index, there are children:

![models-comp-tree](https://user-images.githubusercontent.com/2664090/67947502-14b48180-fbe4-11e9-887e-c44f1cc6ef3d.png)

The redirect after delete still needs fixing, [ForemanModal](https://github.com/theforeman/foreman/pull/7039/) would be a perfect fit for confirmation dialog.

xprazak2 commented on webpack/assets/javascripts/react_app/components/common/table/components/NameCell.js:
```diff
@@ -1,9 +1,10 @@
 const NameCell = ({ active, id, name, controller, children }) =>
   active ? (
-    <a href={`/${controller}/${id}-${name}/edit`}>{children}</a>
+    <Link to={`/${controller}/${id}-${name}/edit`}>{children}</Link>
```
@MariaAga, does this break foreman-tasks?

MariaAga replied:
This is not used in foreman tasks, thanks for asking

jeremylenz commented on app/helpers/form_helper.rb:
```diff
@@ -239,17 +239,24 @@ def soft_limit_warning_block
-  def submit_or_cancel(f, overwrite = false, args = { })
+  def submit_or_cancel(f, overwrite = false, args = { }, &block)
     args[:cancel_path] ||= send("#{controller_name}_path")
+    cancel_btn = block_given? ? yield(args) : link_to(_("Cancel"), args[:cancel_path], :class => "btn btn-default")
```
Interesting way of achieving a React cancel button in an ERB form!  Instead of consuming a `&block` here, would it be simpler to just accept a `:react_cancel_button => true` argument in this `submit_or_cancel` method, and then here in the ternary call `react_cancel_button(args[:cancel_path])` ?

xprazak2 replied:
Using an argument would be simpler, thanks.

jeremylenz commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js:
```diff
@@ -0,0 +1,77 @@
+      onBookmarkClick={handleSearch}
+      toolbarButtons={
+        <Link to="/models/new">
+          <Button bsStyle="primary">{__('Create Model')}</Button>
```
I think this button should be hidden if the user doesn't have `can_create` permission.  Should be straightforward to implement since we already request `include_permissions=true` and I see the `can_create` permission coming back in the API response.

jeremylenz commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPageActions.js:
```diff
@@ -0,0 +1,130 @@
+        order: sortString,
+      }
+    );
+    return onSuccess(response);
```
I feel like it's less readable to define the nested functions `onSuccess` and `onError`, and also doesn't conform to action creators elsewhere.   (I was confused for a moment about why your `dispatch`es above didn't have `await`.)  I feel like it's more clear if all the `try` logic is in your `try`, and all your `catch` logic is in your `catch`.

jeremylenz commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/components/ModelsPageContent.js:
```diff
@@ -0,0 +1,51 @@
+  perPage,
+}) => (
+  <React.Fragment>
+    <div>&nbsp;</div>
```
What is the purpose of this div?

xprazak2 replied:
It adds space between search bar and table, erb layout has it:
![media-nbsp](https://user-images.githubusercontent.com/2664090/68017240-89e08f00-fc96-11e9-8df9-ecc401d972ea.png)

It was actually in the wrong place, so I moved it.

jeremylenz commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -27,17 +28,26 @@ const AppSwitcher = () => (
         }
         currentLocation = child.location;
-        return null;
+        return props.children ? props.children : null;
```
Could be `null || props.children` ... or maybe just `props.children`

jeremylenz reviewed (changes requested):
* When I click an existing model to go to the Edit screen, the page loads correctly now.  But it seems it's not using React Router.  Is the edit form not implemented in React yet?  In order to get the full benefit of React Router, I feel like we need to have both pages implemented in React.  (If it's a follow-up to this PR I'm fine with that.)
* Cancel button is awesome now.  The React tree stays intact and I immediately see the previous page.  :heavy_check_mark:

xprazak2 commented:
Main changes:

* checking for permissions to display button in toolbar
* I removed the nested functions and now logic is directly inside try/catch blocks

Edit and new pages are not in react yet, that is the reason transitioning to these pages is not handled by react-router and I need to pull in the routes for cancel button.

I still need to fix that redirect after delete.

jeremylenz reviewed (changes requested):
Changes look good, thanks!

One thing I missed before-- if I am on the new or edit page, and I click the Hardware Models breadcrumb, I am seeing a blank page.

glekner commented on webpack/assets/javascripts/react_app/components/common/RedirectCancelButton/RedirectCancelButton.js:
```diff
@@ -0,0 +1,23 @@
+import withReactRoutes from '../../../common/withReactRoutes';
+import { translate as __ } from '../../../common/I18n';
+
+const RedirectCancelButton = props => (
```
This is a good temporary idea until all the pages are migrated to the router

glekner commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/index.js:
```diff
@@ -0,0 +1,67 @@
+});
+
+export const initialState = Immutable({
+  total: 0,
```
why not use `withQueryReducer` for the pagination and search fields?

xprazak2 replied:
Because sort order and sort attribute are part of the query when sorting by table column, technically they should be in `withQueryReducer`. That means I would need to extend it with additional state as I do with `withDataReducer`. I would be extending 2 functions in the exact same way to create 2 reducers that handle different part of the same response. But I did not notice any benefits in splitting response payload in 2 different branches of the state tree, I was just dispatching more actions.

glekner commented on webpack/assets/javascripts/react_app/routes/common/PageLayout/PageLayout.js:
```diff
@@ -50,16 +50,15 @@ const PageLayout = ({
+                />
               </div>
             )}
+            &nbsp;
```
why is this needed?

xprazak2 replied:
To be consistent with erb layout and have the same spacing between search bar and table.

https://github.com/theforeman/foreman/pull/6818#discussion_r341311263

glekner commented on webpack/assets/javascripts/react_app/routes/common/reducerHOC/withDataReducer.js:
```diff
@@ -1,14 +1,14 @@
 
-const withDataReducer = controller => (
-  state = initialState,
+const withDataReducer = (controller, additionalState = Immutable({})) => (
```
🚀

glekner commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/components/ModelsPageContent.js:
```diff
@@ -0,0 +1,50 @@
+
+export default withRenderHandler({
+  Component: ModelsPageContent,
+  LoadingComponent: LoadingPage,
```
I Think a custom table skeleton for the Loading component could be really cool

glekner commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/components/ModelsPageContent.js:
```diff
@@ -0,0 +1,50 @@
+      sortOrder={sort.order}
+      getTableItems={fetchAndPush}
+    />
+    <Pagination
```
I noticed a margin between the table and the pagination component, is it intentional?

xprazak2 replied:
No, not intentional, will fix.

glekner commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js:
```diff
@@ -0,0 +1,81 @@
+      searchable
+      searchProps={MODELS_SEARCH_PROPS}
+      searchQuery={search}
+      isLoading={isLoading && hasData}
```
maybe not in the scope of this PR, but a 100ms delay before showing the little spinner could be very nice

glekner commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -27,17 +28,26 @@ const AppSwitcher = () => (
+            (child.location.state &&
+              child.location.state.useTurbolinks &&
+              !window.history.state.turbolinks) ||
+            !child.location.state;
```
can you explain this change

xprazak2 replied:
When navigating to erb page, there might not be a state, but we still want to use turbolinks: 

![no-match](https://user-images.githubusercontent.com/2664090/68308594-69a53b80-00ad-11ea-9857-d900934c9a04.png)

Taken from https://github.com/theforeman/foreman/pull/6604/commits/fcfb89ea446cb7b779b301303d82e412018c62ec

glekner reviewed (changes requested):
This is looking very good @xprazak2 
Left a few comments, I think the breadcrumb issues are interesting to solve and once all the views are in react we could work out a seamless navigation between them.

jeremylenz commented:
Looks like some snapshots need to be updated

xprazak2 commented:
Rebased and updated, now needs https://github.com/theforeman/foreman/pull/7354 to confirm deletion.

jeremylenz commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableSchema.js:
```diff
@@ -21,16 +31,23 @@ import {
+const createModelsTableSchema = (apiCall, by, order, setToDelete) => {
   const sortController = sortControllerFactory(apiCall, by, order);
+
+  const { setModalOpen } = useForemanModal({ id: MODEL_DELETE_MODAL_ID });
```
I think you'll have to find a way to move this Hooks call into a component, so it doesn't break the [Rules of Hooks](https://reactjs.org/docs/hooks-rules.html) (yet another downside of the patternfly-react 3 table API).

jeremylenz commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/components/ModelsPageContent.js:
```diff
@@ -0,0 +1,57 @@
+  page,
+  perPage,
+}) => {
+  const [toDelete, setToDelete] = useState();
```
Prefer passing an initial value to `useState` just to show what the data will look like (otherwise I think it will start out with a state of `undefined`)

jeremylenz commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js:
```diff
@@ -0,0 +1,81 @@
+      isLoading={isLoading && hasData}
+      onSearch={handleSearch}
+      onBookmarkClick={handleSearch}
+      toolbarButtons={canCreate && createBtn}
```
When I'm on the models page and I click Create, I'm still getting the blank screen.

jeremylenz commented on webpack/assets/javascripts/react_app/components/common/table/formatters/nameCellFormatter.js:
```diff
@@ -3,7 +3,7 @@ import NameCell from '../components/NameCell';
 
 const nameCellFormatter = controllerPluralize => (
```
When I click on a name to go to the Edit page, I'm still getting a blank screen

jeremylenz commented on app/views/models/_form.html.erb:
```diff
@@ -4,5 +4,5 @@
   <%= text_f f, :vendor_class, :help_block => _("The class of the machine reported by the Open Boot Prom. This is primarily used by Sparc Solaris builds and can be left blank for other architectures. The value can be determined on Solaris via uname -i|cut -f2 -d,") %>
   <%= textarea_f f, :info, :rows=> 7, :size => "col-md-8", :help_block => _("General useful description, for example this kind of hardware needs a special BIOS setup") %>
-  <%= submit_or_cancel f %>
+  <%= submit_or_cancel(f, false, { :react_cancel_button => true }) %>
```
Not sure how ambitious you wanna be, but it would be awesome if the Submit button could be React also.  That way you could use the mini loader and redirect back to the table fully with React Router.  (At this point it'd probably be better just to re-do the whole form in React)

xprazak2 replied:
I agree it would be nice and I'd like to keep the changes contained to the index page as this has almost 1500 changed lines already.

jeremylenz commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPageActions.js:
```diff
@@ -0,0 +1,116 @@
+
+const checkSort = sort => (sort && sort.by && sort.order ? sort : undefined);
+
+const transformSort = sort => ({ ...sort, by: snakeCase(sort.by) });
```
I feel like these should be moved to a `ModelsPageHelpers.js`

jeremylenz commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPageActions.js:
```diff
@@ -0,0 +1,116 @@
+  });
+};
+
+const buildQuery = (query, state) => {
```
Same for this one

jeremylenz commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -49,7 +50,7 @@ const AppSwitcher = () => {
       handleTurbolinksVisit(location, nextPath);
     }
-    return null;
+    return props.children ? props.children : null;
```
Can this just be `return props.children` ?

jeremylenz reviewed (changes requested):
Thanks @xprazak2, getting closer!  Comments below

jeremylenz commented on webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableSchema.js:
```diff
@@ -10,8 +10,8 @@ import {
-import { useForemanModal } from '../ForemanModal/ForemanModalHooks';
-import { MODEL_DELETE_MODAL_ID } from '../../routes/Models/constants';
+// import { useForemanModal } from '../ForemanModal/ForemanModalHooks';
+// import { MODEL_DELETE_MODAL_ID } from '../../routes/Models/constants';
```
leftover?

jeremylenz commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/components/ModelDeleteModal.js:
```diff
@@ -16,11 +15,8 @@ const ModelDeleteModal = props => {
-        ),
-        onSucc: props.fetchAndPush,
+        message: sprintf(__('Hardware Model %s was successfuly deleted'), name),
+        onSuccess: props.fetchAndPush,
```
Should add `submitBtnProps` here so the button text says "Delete" and the `bsStyle` is "danger"

jeremylenz commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/components/ModelDeleteModal.js:
```diff
@@ -16,11 +15,8 @@ const ModelDeleteModal = props => {
       enforceFocus
       submitProps={{
         url: `/api/v2/models/${id}`,
```
I think `id` is coming in `undefined` here.  When I click submit on the delete confirmation dialog, 
the DELETE request is going to `/api/v2/models/undefined`

and I'm getting an error:
```
Resource model not found by id 'undefined'
```

xprazak2 replied:
Interesting, I did not run into that. The correct `id` is always set when I click the delete button in the table.

jeremylenz commented on webpack/assets/javascripts/react_app/components/ModelsTable/index.js:
```diff
@@ -1 +1,24 @@
+import { useForemanModal } from '../ForemanModal/ForemanModalHooks';
+
+const WrappedModelsTable = props => {
+  const { setModalOpen } = useForemanModal({ id: MODEL_DELETE_MODAL_ID });
```
:+1: :+1:  This is a much better place for the hook

jeremylenz reviewed (changes requested):
Thanks @xprazak2 

* :heavy_check_mark: Blank pages aren't happening any more

Headed the right direction.  A few issues/comments below

xprazak2 commented:
I removed the leftovers. I also changed the style and text for submit button. Unfortunately, I could not reproduce the issue with missing id. Does that happen every time?

jeremylenz commented:
@xprazak2 Not sure what was causing it.  I am no longer able to reproduce it at all.  @MariaAga or @sharvit can you double check and see if you get the [issue](https://github.com/theforeman/foreman/pull/6818#discussion_r372597607)?

jeremylenz reviewed (dismissed):
Per Travis output, there's a snapshot that needs to be updated.  Changes look good, ACK pending Travis and 1 more functional test by someone who's not me.

(To test the PR, I checked it out and then rebased on the latest commit of #7354)

MariaAga commented:
I tried to delete one, the `submitProps` had the right ID but there isn't a `Delete` button on the modal and I get an error in console:
```
backend.js:6 Warning: React does not recognize the `submitProps` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `submitprops` instead. If you accidentally passed it from a parent component, remove it from the DOM element.
    in div (created by CustomModalDialog)
    in CustomModalDialog (created by Modal)
    in Transition (created by Fade)
    in Fade (created by DialogTransition)
    in DialogTransition (created by Modal)
    in RefHolder (created by Modal)
    in div (created by Modal)
    in Portal (created by Modal)
    in Modal (created by Modal)
    in Modal (created by ForemanModal)
    in ForemanModal (created by ConnectedForemanModal)
```

xprazak2 commented:
@MariaAga, you need to rebase on https://github.com/theforeman/foreman/pull/7354, it contains additions to the modal.

MariaAga commented:
@xprazak2 Thanks, works fine now :)

ofedoren reviewed (commented):
Here is my two cents:

 - Is it intentional that the models table has bottom margin in cca 6px? Other tables don't have that, e.g. hosts table.
 - When I create/update a model the notification message of success is shown after I go to a different page only.. This is kinda weird. Deletion works as expected though.

xprazak2 commented:
I fixed the space between pagination and table, the toasts now show properly.

ofedoren reviewed (commented):
Thanks, @xprazak2! Unfortunately, JS is not my field, so can't say about it much, but the page itself works for me just fine :)

sharvit commented on app/helpers/layout_helper.rb:
```diff
@@ -1,6 +1,10 @@
+      layout: layout_data,
+      metadata: app_metadata,
+      toasts: toast_notifiations_data,
+    }.to_json, :flatten_data => true)
```
What is the reason for flattening? The `data` prop is a convention so we can identify it's data loaded by the server on rendering-time.

xprazak2 replied:
So that I do not have to destruct [later]{https://github.com/theforeman/foreman/pull/6818/files#diff-63cd48aa2e35bd947daf1c571e9521fcR9}, reverted.

sharvit commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.scss:
```diff
@@ -0,0 +1,3 @@
+table.table.table-striped {
```
If it's should be active only in the models page, it should be wrapped with the class of the page.

sharvit commented on webpack/assets/javascripts/react_app/Root/ReactApp.js:
```diff
@@ -6,24 +6,28 @@ import { getForemanContext } from '../Root/Context/ForemanContext';
       <Router history={history}>
         <Layout data={layout}>
-          <AppSwitcher />
+          <AppSwitcher toasts={toasts} />
```
It's a bit hacky passing it down to everyone that way, would it make sense adding it to the `ForemanContext`?

sharvit replied:
Also, why would we want to keep that data? Can't we just `dispatch(addToasts(toasts))`?
They are managed in redux anyway

xprazak2 replied:
`dispatch(addToasts(toasts))` will not work here as they are not displayed properly during page transition.

sharvit replied:
Do you still need to pass down the `toasts` prop?
It's in the context now and globally accessible, I guess you can revert some changes from the `AppSwitcher` component.

sharvit commented on app/registries/menu/loader.rb:
```diff
@@ -56,8 +56,12 @@ def self.load
+              :caption => N_('Hardware Models'),
+              :url => '/models',
+              :url_hash => { :controller => 'api/v2/models', :action => 'index' },
+              :turbolinks => false
```
Do we still need it after we dropped turbolinks?

xprazak2 replied:
No longer needed, removing.

sharvit commented on webpack/assets/javascripts/react_app/common/__snapshots__/withReactRoutes.test.js.snap:
```diff
@@ -0,0 +1,32 @@
+
+exports[`withReactRoutes should render and pass props: should render and pass props 1`] = `
+<Router
+  history={
```
I afraid the interface of the `history` object is going to change between versions.
Can we mock a nicer object instead?

sharvit commented on webpack/assets/javascripts/react_app/common/urlHelpers.js:
```diff
@@ -36,13 +36,29 @@ export const getURIperPage = () => Number(getURI().query(true).per_page);
+  if (!sortString) {
+    return {};
+  }
+  const split = sortString.split(' ');
```
Would be much nicer as
```js
const [by, order] = sortString.split(' ');

return { by, order };
```

sharvit commented on webpack/assets/javascripts/react_app/common/withReactRoutes.js:
```diff
@@ -0,0 +1,15 @@
+import history from '../history';
+import AppSwitcher from '../routes';
+
+const withReactRoutes = Component => props => (
```
IMO we are going to see more uses of `withReactRoutes`, can you document it in the storybook? I guess MDX style would work better

sharvit commented on webpack/assets/javascripts/react_app/components/common/RedirectCancelButton/__snapshots__/RedirectCancelButton.test.js.snap:
```diff
@@ -0,0 +1,32 @@
+
+exports[`RedirectCancelButton renders correctly 1`] = `
+<Router
+  history={
```
Can you mock the history object to get a cleaner span?

sharvit commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPageActions.js:
```diff
@@ -0,0 +1,88 @@
+    const sortString =
+      sort && Object.keys(sort).length > 0 ? `${sort.by} ${sort.order}` : '';
+
+    const { data } = await API.get(
```
Would it make sense to migrate to? http://theforeman.github.io/foreman/?path=/docs/introduction-api-middleware-usage--page

IMO it would simplify a lot in the state management

sharvit commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPageActions.js:
```diff
@@ -0,0 +1,88 @@
+  }
+};
+
+export const fetchAndPush = (params = {}) => (dispatch, getState) => {
```
Can't we do only push and do the fetch on location change event?

xprazak2 replied:
I am not sure what you mean, this follows the convention established by the audits page.

sharvit reviewed (changes requested):
Thanks @xprazak2 👍

xprazak2 commented:
I rebased and switched to the middleware API as suggested.

sharvit commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -15,10 +16,10 @@ const AppSwitcher = () => {
   };
 
-  const handleRoute = (Component, props) => {
+  const handleRoute = (Component, componentProps) => {
```
Do we still need those changes after migrating to use context?

sharvit reviewed (changes requested):
Thanks @xprazak2 👍

Commit:
Refs #26772 - Use toasts from context

sharvit commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -37,12 +38,20 @@ const AppSwitcher = () => {
   );
 };
 
+AppSwitcher.propTypes = {
```
There are no props used by the AppSwitcher now, can remove the prop-types and default-props

xprazak2 replied:
Thanks, I removed the prop types.

sharvit reviewed (changes requested):
Thanks 👍

xprazak2 commented:
Closing in favour of https://github.com/theforeman/foreman/pull/7775, thanks!

sharvit commented:
Thanks @xprazak2, notice the linting issue

sharvit commented:
Thanks @xprazak2, have you noticed that warning?
```
Warning: Render methods should be a pure function of props and state;
triggering nested component updates from render is not allowed.
If necessary, trigger nested updates in componentDidUpdate.

Check the render method of Context.Consumer. 
```

I am getting it when navigating from the dashboard to the hw models page or the audit page.

sharvit commented:
> Thanks @xprazak2, have you noticed that warning?
> 
> ```
> Warning: Render methods should be a pure function of props and state;
> triggering nested component updates from render is not allowed.
> If necessary, trigger nested updates in componentDidUpdate.
> 
> Check the render method of Context.Consumer. 
> ```
> 
> I am getting it when navigating from the dashboard to the hw models page or the audit page.

I have noticed that in `develop` as well, opened an issue: https://projects.theforeman.org/issues/30230

sharvit reviewed (approved):
Thanks @xprazak2 looks good and works as expected 👍
